### PR TITLE
chore(website): auto-generate type checked rule notice in rule docs

### LINKED ISSUE #7918: Docs: Auto-add a performance warning to typed rules' "When Not To Use It"
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

Several typed rules have a similar kind of _"performance is slower with typed rules"_ notice in their _`## When Not To Use It`_. Examples:

https://github.com/typescript-eslint/typescript-eslint/blob/3cc3d0eff8e9d574d35c05c97b607a8a6ba59d9b/packages/eslint-plugin/docs/rules/no-unnecessary-condition.md?plain=1#L98

https://github.com/typescript-eslint/typescript-eslint/blob/3cc3d0eff8e9d574d35c05c97b607a8a6ba59d9b/packages/eslint-plugin/docs/rules/prefer-reduce-type-parameter.md?plain=1#L58

How about we standardize this by adding that automatically in rule docs generation (https://github.com/typescript-eslint/typescript-eslint/blob/3cc3d0eff8e9d574d35c05c97b607a8a6ba59d9b/packages/website/plugins/generated-rule-docs.ts)?

### Affected URL(s)

https://typescript-eslint.io/rules/*

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7918
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds the following notice to the _When Not To Use It_ section of rule docs, auto-generating that h2 if it doesn't yet exist:

> Type checked lint rules are more powerful than traditional lint rules, but also require configuring [type checked linting](http://localhost:3000/linting/typed-linting). See [Performance Troubleshooting](http://localhost:3000/troubleshooting/performance-troubleshooting) if you experience performance degredations after enabling type checked rules.

Bikeshedding welcome on that message.

I also got fed up trying to work with the >500 lines of `generated-rule-docs.ts`, so I enacted two refactors:

1. Splitting up the major parts of its logic into individual files/functions with a few shared utilities. Now `generated-rule-docs/index.ts` has IMO a much more readable flow.
2. Adding a `spliceChildrenAndAdjustHeadings` function that adjusts all headings' indices.  Previously the code mostly manually did it.

### CODE REVIEW

File: packages/website/plugins/generated-rule-docs/spliceWhenNotToUseIt.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import type * as mdast from 'mdast';
+import type * as unist from 'unist';
+
+import type { RequiredHeadingIndices } from './requiredHeadings';
+import { spliceChildrenAndAdjustHeadings } from './splicing';
+import type { RuleMetaDataWithDocs } from './types';
+
+export function spliceWhenNotToUseIt(
```
Comment (Josh-Cena) [inline packages/website/plugins/generated-rule-docs/spliceWhenNotToUseIt.ts:8]: Can we not call functions `spliceX`, `unshiftX`, or `pushX`? They both function as `insert` and which one is used is an implementation detail.

File: packages/website/plugins/generated-rule-docs/nodes.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import type * as mdast from 'mdast';
+import type * as unist from 'unist';
+
+export function nodeIsCode(node: unist.Node): node is mdast.Code {
```
Comment (Josh-Cena) [inline packages/website/plugins/generated-rule-docs/nodes.ts:4]: I suggest moving all non-tree-modifying helper functions to the same file.

**Review by Josh-Cena — COMMENTED** [review]
Just some tiny nits

File: packages/website/plugins/generated-rule-docs/spliceWhenNotToUseIt.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/website/plugins/generated-rule-docs/spliceWhenNotToUseIt.ts:8]: 😍 I love this comment so much. Yes!

File: packages/website/plugins/generated-rule-docs/nodes.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [reply packages/website/plugins/generated-rule-docs/nodes.ts:4]: Sorry, not sure how to parse this - which helper functions are you thinking of?

**Comment by bradzacher** [comment]
Perhaps it would be good to add this as a subsection if there already is a "When not to use it" section?
That way our hand-crafted text sits more prominently and is separated from the auto-generated text.

EG `await-thenable`:

screenshot of docs

It's hard to tell that there are two separate points there and additionally the hand-crafted point is placed second.

File: packages/website/plugins/generated-rule-docs/nodes.ts (hunk 2 above)
Comment (Josh-Cena) [reply packages/website/plugins/generated-rule-docs/nodes.ts:4]: I meant these files: nodes.ts, requiredHeadings.ts, spliceChildrenAndAdjustHeadings.ts, strings.ts, types.ts. These are all small utilities and I don't entertain the idea of splitting them into 5 files well.

File: packages/website/plugins/generated-rule-docs/index.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  nodeIsParent,
+} from './utils';
+
+export const generatedRuleDocs: Plugin = () => {
+  return (root, file) => {
+    if (!nodeIsParent(root) || !isVFileWithStem(file)) {
+      return;
+    }
+
+    // const rule = file.stem in pluginRules ? pluginRules[file.stem] : undefined;
```
Comment (JoshuaKGoldberg) [inline packages/website/plugins/generated-rule-docs/index.ts:26]:
```suggestion
```

**Review by bradzacher — APPROVED** [review]

**Pull request merged**